src/compositemark/boxplot.ts

```diff
@@ -165,86 +160,193 @@ export function normalizeBoxPlot(spec: GenericUnitSpec<Encoding<Field>, BOXPLOT
   const {mark: mark, encoding: encoding, ...outerSpec} = spec;
 
   let discreteAxisFieldDef: PositionFieldDef<Field>;
   let continuousAxisChannelDef: PositionFieldDef<Field>;
   let discreteAxis;
   let continuousAxis;
 
-  // FIXME: refactor code such that we don't have to do this casting
-  // We can cast here as we already check from outside that both x and y are FieldDef
-  const xDef = encoding.x as FieldDef<Field>;
-  const yDef = encoding.y as FieldDef<Field>;
-
+  const is2D = encoding.x && encoding.y;
 
-  if (orient === 'vertical') {
-    discreteAxis = 'x';
-    continuousAxis = 'y';
-    continuousAxisChannelDef = yDef;
-    discreteAxisFieldDef = xDef;
+  if (!is2D) {
+    if (orient === 'horizontal') {
+      continuousAxis = 'x';
+      if (isFieldDef(encoding.x)) {
+        continuousAxisChannelDef = encoding.x;
+      }
+    } else {
+      continuousAxis = 'y';
+      if (isFieldDef(encoding.y)) {
+        continuousAxisChannelDef = encoding.y;
+      }
+    }
   } else {
-    discreteAxis = 'y';
-    continuousAxis = 'x';
-    continuousAxisChannelDef = xDef;
-    discreteAxisFieldDef = yDef;
-  }
+    if (isFieldDef(encoding.x) && isFieldDef(encoding.y)) {
+      if (orient === 'vertical') {
+        discreteAxis = 'x';
+        continuousAxis = 'y';
+        continuousAxisChannelDef = encoding.y;
+        discreteAxisFieldDef = encoding.x;
+      } else {
+        discreteAxis = 'y';
+        continuousAxis = 'x';
+        continuousAxisChannelDef = encoding.x;
+        discreteAxisFieldDef = encoding.y;
+      }
+    }
 
-  if (continuousAxisChannelDef && continuousAxisChannelDef.aggregate) {
-    const {aggregate: aggregate, ...continuousAxisWithoutAggregate} = continuousAxisChannelDef;
-    if (aggregate !== BOXPLOT) {
-      throw new Error(`Continuous axis should not have customized aggregation function ${aggregate}`);
+    if (continuousAxisChannelDef && continuousAxisChannelDef.aggregate) {
+      const {aggregate: aggregate, ...continuousAxisWithoutAggregate} = continuousAxisChannelDef;
+      if (aggregate !== BOXPLOT) {
+        throw new Error(`Continuous axis should not have customized aggregation function ${aggregate}`);
+      }
+      continuousAxisChannelDef = continuousAxisWithoutAggregate;
     }
-    continuousAxisChannelDef = continuousAxisWithoutAggregate;
   }
 
   return {
     discreteAxisFieldDef: discreteAxisFieldDef,
     continuousAxisChannelDef: continuousAxisChannelDef,
     discreteAxis: discreteAxis,
-    continuousAxis: continuousAxis
+    continuousAxis: continuousAxis,
+    is1D: !is2D
   };
 }
+
+export function boxTransform(encoding: Encoding<Field>, discreteAxisFieldDef: PositionFieldDef<Field>, continuousAxisChannelDef: PositionFieldDef<Field>, kIQRScalar: 'min-max' | number, is1D: boolean) {
+  const isMinMax = kIQRScalar === undefined;
+  const groupbyDef = boxTransformGroupby(discreteAxisFieldDef, encoding);
```

Comment:
Since `groupbyDef` is never used in the chunk below, maybe moving it to right above where you assign it would make the code easier to read. 

Also, rename to `groupBy`.